compiler/frontend/src/hir.rs

```diff
@@ -133,71 +134,125 @@ impl Body {
 #[derive(Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
 pub struct Id {
     pub module: Module,
-    pub keys: Vec<IdKey>,
+    pub keys: IdPath,
 }
+#[derive(Clone, Eq, From, Hash, Ord, PartialEq, PartialOrd, Debug)]
+pub struct IdPath(String);
 #[derive(Clone, Eq, From, Hash, Ord, PartialEq, PartialOrd)]
 pub enum IdKey {
     Named { name: String, disambiguator: usize },
     Positional(usize),
 }
+impl IdPath {
+    const fn empty() -> Self {
+        Self(String::new())
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn push(&self, key: IdKey) -> Self {
+        if self.is_empty() {
+            Self::from(key)
+        } else {
+            Self(format!("{}:{key}", self.0))
+        }
+    }
+
+    #[must_use]
+    pub fn last_as_string(&self) -> Option<&str> {
+        self.0.rfind(':').map(|i| &self.0[(i + ':'.len_utf8())..])
+    }
+
+    fn drop_last(&self) -> Option<Self> {
+        self.0.rfind(':').map(|i| Self(self.0[..i].to_string()))
+    }
+}
+impl From<IdKey> for IdPath {
+    fn from(value: IdKey) -> Self {
+        Self(format!("{value}"))
+    }
+}
+impl From<Vec<IdKey>> for IdPath {
+    fn from(value: Vec<IdKey>) -> Self {
+        Self(value.iter().map(|it| format!("{it}")).join(":"))
```

Comment:
`join(…)` formats the items automatically

```suggestion
        Self(value.iter().join(":"))
```